### DISCUSSION TITLE
Passport.js authenticate issue with two simultaneous sessions : "Cannot set headers after they are sent to the client"

### DISCUSSION THREAD

**Discussion by xmonvoisin** [jaredhanson/passport#995] [Help]
Hi,

I have issue with this code when testing using 2 simultaneous users (Playwright tests, using 2 workers) :

`passport.authenticate(strategy, {}, { failureRedirect: `/login`, state: redirect_url }), (err, user, info) => {
                    if (err || !user) {
                        return next(err || info);
                    }
                    req.logIn(user, err => {
                        if (err) {
                            return next(err);
                        }                      
                        if (loginpopup) {
                            ...
                        }
                        else {
                            return res.redirect('/');
                        }
                    });
                })(req, res, next);
`
I am receiving this error: 

`Error: Cannot set headers after they are sent to the client ...`

It seems to come from the "res.redirect('/')"
From various posts on stackoverflow, I kind of understand that the issue come from the fact that I do a redirect inside the "authenticate" callback and that redirect interfer with passport's redirect. But I don't know how to solve this.

Could you please help me understand and fix this issue ?
Thanks you

**Comment by fergus99** [comment]
I had a very similar issue.  I fixed it by replacing "passport.authenticate(strategy..." with "passport.authenticate('strategyName'...".
I think this is down to a bug in Passport whereby the authenticate function creates the strategy using Object.create() on the prototype only in the case of the strategy being passed as a string.  Alternatively you could use passport.authenticate(Object.create(strategyName)...".

**Reply by BeataKr** [reply]
Thank you. Seems to be a good fix.

**Reply by mhoberg9** [reply]
Just came across this error while requesting data from multiple endpoints on page load. 

Thanks for sharing, this fixed my header error.

Followed the steps from the original documentation (https://www.passportjs.org/concepts/authentication/strategies - #Register) and named my custom strategy first and then used the string for authenticate().

- passport.use("jwt", jwtStrategy);
- passport.authenticate("jwt", ... )

**Comment by usualdesigner** [comment]
'Cannot set headers after they are sent' basically always means res.something() is getting called twice on the same request somewhere in that flow. With two simultaneous sessions this could be a race where both requests hit the same handler and something isn't guarding against being called twice. I'd add `if (res.headersSent) return` at the top of anywhere that sends, and log a stack trace on the second attempt to find exactly which line double-fires.
